src/sentry/tasks/assemble.py

```diff
@@ -452,8 +456,49 @@ def close(self):
         self.archive.close()
 
     def post_assemble(self):
-        with metrics.timer("tasks.assemble.artifact_bundle"):
-            self._create_artifact_bundle()
+        if features.has("organizations:sourcemaps-bundle-flat-file-indexing", self.organization):
+            with metrics.timer("tasks.assemble.artifact_bundle_new"):
+                self._create_artifact_bundle_new()
+        else:
+            with metrics.timer("tasks.assemble.artifact_bundle"):
+                self._create_artifact_bundle()
+
+    def _create_artifact_bundle_new(self):
+        if not self.projects_ids:
+            raise AssembleArtifactsError("uploading a bundle requires at least a project")
+
+        # We want to give precedence to the request fields and only if they are unset fallback to the manifest's
+        # contents.
+        self.release = self.release or self.archive.manifest.get("release")
+        self.dist = self.dist or self.archive.manifest.get("dist")
+
+        # We don't allow an upload without debug ids and release.
+        if not self.archive.has_debug_ids() and not self.release:
+            raise AssembleArtifactsError(
+                "uploading a bundle without debug ids or release is prohibited"
+            )
+
+        # We take a snapshot in time in order to have consistent values in the database.
+        date_snapshot = timezone.now()
+
+        bundle_id = self.archive.extract_bundle_id()
+
+        # We want to build all the required parameters for indexing.
+        artifact_bundle, created = self._bind_or_create_artifact_bundle(
+            bundle_id=bundle_id, date_added=date_snapshot
+        )
+        identifier = FlatFileIdentifier(
+            # For now, we just take the first project since Sentry CLI uploads one project at a time.
+            project_id=self.projects_ids[0],
+            release=self.release or NULL_STRING,
+            dist=self.dist or NULL_STRING,
+        )
+
+        # We want to asynchronously compute the index, in order to speed up the process and have the possibility to
+        # reschedule the task in case of contention.
+        index_artifact_bundle.apply_async(
```

Comment:
why is this async suddenly? The current indexing that happens in the database is fully sync. We only finish `assemble` when the index has been written.